File: android/src/enabled/java/expo/modules/truecaller/ExpoTruecallerModule.kt
package expo.modules.truecaller

import android.content.Context
import androidx.activity.ComponentActivity
import androidx.activity.result.contract.ActivityResultContracts
import com.truecaller.android.sdk.oAuth.CodeVerifierUtil
import com.truecaller.android.sdk.oAuth.OAuthThemeOptions
import com.truecaller.android.sdk.oAuth.TcOAuthCallback
import com.truecaller.android.sdk.oAuth.TcOAuthData
import com.truecaller.android.sdk.oAuth.TcOAuthError
import com.truecaller.android.sdk.oAuth.TcSdk
import com.truecaller.android.sdk.oAuth.TcSdkOptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CompletableDeferred
import java.math.BigInteger
import java.security.SecureRandom
import java.util.Locale

class ExpoTruecallerModule : Module() {
  @Volatile private var isSdkReady = false
  private var sdkInitResult: CompletableDeferred<Unit>? = null
  private var stateRequested: String? = null
  private var codeVerifier: String? = null

  private val tcOAuthCallback = object : TcOAuthCallback {
    override fun onSuccess(tcOAuthData: TcOAuthData) {
      if (tcOAuthData.state != stateRequested) {
        sendEvent(
          "onVerificationResult",
          mapOf(
            "status" to "failure",
            "error" to "UNKNOWN",
            "errorMessage" to "OAuth state mismatch - the response didn't match the request " +
              "this module made, so it was rejected as a possible forged/replayed response."
          )
        )
        return
      }

      sendEvent(
        "onVerificationResult",
        mapOf(
          "status" to "success",
          "platform" to "android",
          "authorizationCode" to tcOAuthData.authorizationCode,
          "codeVerifier" to codeVerifier,
          "scopesGranted" to tcOAuthData.scopesGranted,
          "simState" to tcOAuthData.simState,
          "deviceCode" to tcOAuthData.deviceCode
        )
      )
    }

    override fun onFailure(tcOAuthError: TcOAuthError) {
      if (!isSdkReady) {
        TcSdk.clear()
        sdkInitResult?.complete(Unit)
        sdkInitResult = null
      }

      sendEvent(
        "onVerificationResult",
        mapOf(
          "status" to "failure",
          "error" to mapTcErrorCode(tcOAuthError.errorCode),
          "errorMessage" to tcOAuthError.errorMessage
        )
      )
    }

    override fun onVerificationRequired(tcOAuthError: TcOAuthError?) {
      sendEvent(
        "onVerificationResult",
        mapOf(
          "status" to "dismissed",
          "errorMessage" to "Truecaller can't verify this user via the one-tap flow (e.g. " +
            "they're not a Truecaller user, or their SIM can't be verified this way) - fall " +
            "back to your app's other verification method."
        )
      )
    }

    override fun onSdkReady() {
      isSdkReady = true
      sdkInitResult?.complete(Unit)
    }
  }

  private suspend fun ensureSdkInitialized(context: Context, options: Map<String, Any?>? = null): Boolean {
    if (isSdkReady) return true

    val deferred = sdkInitResult ?: CompletableDeferred<Unit>().also { newDeferred ->
      sdkInitResult = newDeferred
      try {
        val builder = TcSdkOptions.Builder(context, tcOAuthCallback)
        applyCustomization(builder, options)
        TcSdk.initAsync(builder.build())
      } catch (e: Exception) {
        sdkInitResult = null
        sendEvent(
          "onVerificationResult",
          mapOf(
            "status" to "failure",
            "error" to "INTERNAL_ERROR",
            "errorMessage" to (e.message ?: "Truecaller SDK failed to initialize.")
          )
        )
        newDeferred.completeExceptionally(e)
      }
    }

    return try {
      deferred.await()
      isSdkReady
    } catch (e: Exception) {
      false
    }
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoTruecaller")

    Events("onVerificationResult")

    AsyncFunction("initialize") Coroutine { options: Map<String, Any?>? ->
      val context = appContext.reactContext
        ?: throw IllegalStateException("No context available to initialize Truecaller SDK")
      ensureSdkInitialized(context, options)
      Unit
    }

    AsyncFunction("isTruecallerUsable").Coroutine<Boolean> {
      val activity = appContext.currentActivity ?: return@Coroutine false
      if (!ensureSdkInitialized(activity)) return@Coroutine false
      TcSdk.getInstance().isOAuthFlowUsable
    }

    AsyncFunction("requestVerification") Coroutine { options: Map<String, Any?>? ->
      val activity = appContext.currentActivity
        ?: throw IllegalStateException("No activity available to launch Truecaller verification")

      if (!ensureSdkInitialized(activity)) return@Coroutine

      if (!TcSdk.getInstance().isOAuthFlowUsable) {
        sendEvent(
          "onVerificationResult",
          mapOf(
            "status" to "failure",
            "error" to "NOT_INSTALLED",
            "errorMessage" to "Truecaller app is not installed, not signed in, or otherwise unusable on this device."
          )
        )
        return@Coroutine
      }

      val newState = BigInteger(130, SecureRandom()).toString(32)
      stateRequested = newState
      TcSdk.getInstance().setOAuthState(newState)

      @Suppress("UNCHECKED_CAST")
      val requestedScopes = (options?.get("scopes") as? List<String>)
        ?.takeIf { it.isNotEmpty() }
        ?.toTypedArray()
        ?: arrayOf("profile", "phone")
      TcSdk.getInstance().setOAuthScopes(requestedScopes)

      val newCodeVerifier = CodeVerifierUtil.generateRandomCodeVerifier()
      codeVerifier = newCodeVerifier
      val codeChallenge = CodeVerifierUtil.getCodeChallenge(newCodeVerifier)

      if (codeChallenge == null) {
        sendEvent(
          "onVerificationResult",
          mapOf(
            "status" to "failure",
            "error" to "UNSUPPORTED_OS_VERSION",
            "errorMessage" to "This device doesn't support the SHA-256/Base64 operations PKCE requires."
          )
        )
        return@Coroutine
      }

      TcSdk.getInstance().setCodeChallenge(codeChallenge)

      (options?.get("theme") as? String)?.let {
        val theme = if (it == "dark") OAuthThemeOptions.DARK else OAuthThemeOptions.LIGHT
        TcSdk.getInstance().setTheme(theme)
      }
      (options?.get("language") as? String)?.let {
        TcSdk.getInstance().setLocale(Locale.forLanguageTag(it))
      }

      val launcher = (activity as ComponentActivity).activityResultRegistry.register(
        "TruecallerOAuth-${System.currentTimeMillis()}",
        ActivityResultContracts.StartActivityForResult()
      ) { /* no-op - see OnActivityResult below */ }

      TcSdk.getInstance().getAuthorizationCode(activity, launcher)
    }

    AsyncFunction("clearCredentials") {
      TcSdk.clear()
      isSdkReady = false
      sdkInitResult = null
    }

    OnActivityResult { activity, payload ->
      TcSdk.getInstance().onActivityResultObtained(activity, payload.resultCode, payload.data)
    }
  }
}

File: ios/Enabled/ExpoTruecallerModule.swift
import ExpoModulesCore
import TrueSDK

private final class TruecallerDelegateProxy: NSObject, TCTrueSDKDelegate {
  weak var module: ExpoTruecallerModule?

  func didReceive(_ profile: TCTrueProfile) {
    module?.handleDidReceive(profile)
  }

  func didFailToReceiveTrueProfileWithError(_ error: TCError) {
    module?.handleDidFailToReceiveTrueProfile(error)
  }

  @objc(willRequestProfileWithNonce:)
  func willRequestProfile(withNonce nonce: String) {
    module?.handleWillRequestProfile(nonce: nonce)
  }

  @objc(didReceiveTrueProfileResponse:)
  func didReceiveProfileResponse(_ profileResponse: TCTrueProfileResponse) {
    module?.handleDidReceiveTrueProfileResponse(profileResponse)
  }
}

public class ExpoTruecallerModule: Module {
  private var isInitialized = false
  private let truecallerDelegate = TruecallerDelegateProxy()

  private var expectedNonce: String?
  private var verificationMaterial: [String: Any]?
  private var nonceMismatchDetected = false

  public func definition() -> ModuleDefinition {
    Name("ExpoTruecaller")

    Events("onVerificationResult")

    AsyncFunction("initialize") { (_ options: [String: Any?]?) in
      let appKey = Bundle.main.object(forInfoDictionaryKey: "TruecallerAppKey") as? String ?? ""
      let appLink = Bundle.main.object(forInfoDictionaryKey: "TruecallerAppLink") as? String ?? ""

      guard !appKey.isEmpty, !appLink.isEmpty else {
        return
      }

      let manager = TCTrueSDK.sharedManager()
      manager.setup(withAppKey: appKey, appLink: appLink)
      self.truecallerDelegate.module = self
      manager.delegate = self.truecallerDelegate
      self.isInitialized = true
    }.runOnQueue(.main)

    AsyncFunction("isTruecallerUsable") { () -> Bool in
      self.isInitialized && TCTrueSDK.sharedManager().isSupported()
    }.runOnQueue(.main)

    AsyncFunction("requestVerification") { (_ options: [String: Any?]?) in
      guard self.isInitialized else {
        self.sendEvent("onVerificationResult", [
          "status": "failure",
          "error": "INTERNAL_ERROR",
          "errorMessage": "Truecaller SDK was never initialized - call initialize() first.",
        ])
        return
      }

      let manager = TCTrueSDK.sharedManager()
      guard manager.isSupported() else {
        self.sendEvent("onVerificationResult", [
          "status": "failure",
          "error": "NOT_INSTALLED",
          "errorMessage": "Truecaller app is not installed or otherwise unusable on this device.",
        ])
        return
      }

      if let language = options?["language"] as? String {
        manager.locale = language
      }

      manager.requestTrueProfile()
    }.runOnQueue(.main)

    AsyncFunction("clearCredentials") { () in
      TCTrueSDK.sharedManager().delegate = nil
      self.truecallerDelegate.module = nil
      self.isInitialized = false
    }.runOnQueue(.main)

    OnDestroy {
      TCTrueSDK.sharedManager().delegate = nil
      self.truecallerDelegate.module = nil
      self.isInitialized = false
    }
  }

  fileprivate func handleWillRequestProfile(nonce: String) {
    expectedNonce = nonce
    verificationMaterial = nil
    nonceMismatchDetected = false
  }

  fileprivate func handleDidReceiveTrueProfileResponse(_ response: TCTrueProfileResponse) {
    if let expected = expectedNonce, let received = response.requestNonce, expected != received {
      nonceMismatchDetected = true
      return
    }

    verificationMaterial = [
      "payload": response.payload as Any,
      "signature": response.signature as Any,
      "signatureAlgorithm": response.signatureAlgorithm as Any,
    ]
  }

  fileprivate func handleDidReceive(_ profile: TCTrueProfile) {
    defer {
      expectedNonce = nil
      verificationMaterial = nil
      nonceMismatchDetected = false
    }

    if nonceMismatchDetected {
      sendEvent("onVerificationResult", [
        "status": "failure",
        "error": "UNKNOWN",
        "errorMessage": "Request nonce mismatch - the response didn't match the request this " +
          "module made, so it was rejected as a possible forged/replayed response.",
      ])
      return
    }

    var gender: String?
    switch profile.gender {
    case .male: gender = "male"
    case .female: gender = "female"
    default: gender = nil
    }

    var profileDict: [String: Any] = [
      "firstName": profile.firstName as Any,
      "lastName": profile.lastName as Any,
      "phoneNumber": profile.phoneNumber as Any,
      "countryCode": profile.countryCode as Any,
      "email": profile.email as Any,
      "gender": gender as Any,
      "avatarUrl": profile.avatarURL as Any,
      "city": profile.city as Any,
      "isVerified": profile.isVerified,
    ]

    if let verificationMaterial {
      profileDict["verification"] = verificationMaterial
    }

    sendEvent("onVerificationResult", [
      "status": "success",
      "platform": "ios",
      "profile": profileDict,
    ])
  }

  fileprivate func handleDidFailToReceiveTrueProfile(_ error: TCError) {
    sendEvent("onVerificationResult", [
      "status": "failure",
      "error": mapTcErrorCode(error.code),
      "errorMessage": error.localizedDescription,
    ])
  }
}
